File: src/x/crazyjvm/command/s/Invoker.scala
package x.crazyjvm.command.s

/**
 * Created by chenchao on 15/10/13.
 */

object Invoker extends App{
  private val history: Seq[() => Unit] = Seq.empty

  def invoke(command: => Unit) { // by-name parameter
    command
    history :+ command _
  }

  Invoker.invoke(println("foo"))

  Invoker.invoke {
    println("bar 1")
    println("bar 2")
  }
}

File: src/x/crazyjvm/command/j/Invoker.java
package x.crazyjvm.command.j;

import java.util.List;
import java.util.ArrayList;

/**
 * Created by chenchao on 15/10/13.
 */
public class Invoker {
    private final List<Runnable> history = new ArrayList<Runnable>();

    void invoke(Runnable command) {
        command.run();
        history.add(command);
    }

    public static void main(String[] args) {
        Invoker invoker = new Invoker();
        invoker.invoke(new PrintCommand("foo"));
        invoker.invoke(new PrintCommand("bar"));
    }
}

